hoomd/GSDDumpWriter.cc

```diff
@@ -1110,6 +1015,482 @@ void GSDDumpWriter::populateNonDefault()
+
+    // attributes
+    frame.particle_data.type.resize(0);
+    frame.particle_data.mass.resize(0);
+    frame.particle_data.charge.resize(0);
+    frame.particle_data.diameter.resize(0);
+    frame.particle_data.body.resize(0);
+    frame.particle_data.inertia.resize(0);
+
+    // momenta
+    frame.particle_data.vel.resize(0);
+    frame.particle_data.angmom.resize(0);
+    frame.particle_data.image.resize(0);
+
+    ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(), access_location::host, access_mode::read);
+
+    if (N > 0)
+        {
+        ArrayHandle<unsigned int> h_tag(m_pdata->getTags(),
+                                        access_location::host,
+                                        access_mode::read);
+
+        m_index.resize(0);
+
+        for (unsigned int group_tag_index = 0; group_tag_index < N; group_tag_index++)
+            {
+            unsigned int tag = m_group->getMemberTag(group_tag_index);
+            unsigned int index = h_rtag.data[tag];
+            if (index >= m_pdata->getN())
+                {
+                continue;
+                }
+
+            frame.particle_tags.push_back(h_tag.data[index]);
+            m_index.push_back(index);
+            }
+        }
+
+    if (N > 0
+        && (m_particle_dynamic[gsd_flag::position] || m_particle_dynamic[gsd_flag::type]
+            || m_particle_dynamic[gsd_flag::image] || m_nframes == 0))
+        {
+        ArrayHandle<Scalar4> h_postype(m_pdata->getPositions(),
+                                       access_location::host,
+                                       access_mode::read);
+        ArrayHandle<int3> h_image(m_pdata->getImages(), access_location::host, access_mode::read);
+
+        if (m_particle_dynamic[gsd_flag::position] || m_nframes == 0)
+            {
+            frame.particle_data_present[gsd_flag::position] = true;
+            }
+        if (m_particle_dynamic[gsd_flag::image] || m_nframes == 0)
+            {
+            frame.particle_data_present[gsd_flag::image] = true;
+            }
+        if (m_particle_dynamic[gsd_flag::type] || m_nframes == 0)
+            {
+            frame.particle_data_present[gsd_flag::type] = true;
+            }
+
+        for (unsigned int index : m_index)
+            {
+            vec3<Scalar> position
+                = vec3<Scalar>(h_postype.data[index]) - vec3<Scalar>(m_pdata->getOrigin());
+            unsigned int type = __scalar_as_int(h_postype.data[index].w);
+            int3 image = make_int3(0, 0, 0);
+
+            if (m_particle_dynamic[gsd_flag::image] || m_nframes == 0)
+                {
+                image = h_image.data[index];
+                }
+
+            frame.global_box.wrap(position, image);
+
+            if (m_particle_dynamic[gsd_flag::position] || m_nframes == 0)
+                {
+                if (position.x != Scalar(0.0) || position.y != Scalar(0.0)
+                    || position.z != Scalar(0.0))
```

Comment:
```suggestion
                if (position != vec3<Scalar>{0.0, 0.0, 0.0})
```
Is there a reason not to use the overloaded `==`?

Reply:
I forget that these exist.